### PR TITLE
Prepare training model using TensorFlow 1.x

### PR DESCRIPTION
**Pull request by utilForever**
This revision includes:
- Create Python script files for training model using TensorFlow 1.x (#358)
- To-do list after it is merged
    - Refactor `model.py` and `train.py`
    - Create Python script files using different frameworks
        - TensorFlow 2.x
        - PyTorch
    - Refactor RosettaTorch structure to use `tiny-dnn` (#357)

### CODE REVIEW

File: Extensions/RosettaTraining/Scripts/TensorFlow/train.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+import argparse
+import numpy
+import random
+import os
+import shutil
+import tempfile
+
+import tensorflow as tf
+from tensorflow.python.tools import freeze_graph
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/train.py]: Codacy Issue found: [Unused freeze_graph imported from tensorflow.python.tools](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 18
           

Complexity increasing per file
==============================
- Extensions/RosettaTraining/Scripts/TensorFlow/train.py  3
- Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py  6
- Extensions/RosettaTraining/Scripts/TensorFlow/model.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929&bid=15844195)

File: Extensions/RosettaTraining/Scripts/TensorFlow/train.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+  export_dir = tempfile.mkdtemp()
+  saved_model = estimator.export_savedmodel(
+      export_dir,
+      tf.estimator.export.build_raw_serving_input_receiver_fn(feature))
+
+  graph_def = saved_model_utils.get_meta_graph_def(
+      saved_model, tag_constants.SERVING).graph_def
+
+  _ = importer.import_graph_def(graph_def, name="")
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/train.py]: Codacy Issue found: [local variable '_' is assigned to but never used](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/train.py (hunk 1 above)
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/train.py]: Codacy Issue found: ['tensorflow.python.tools.freeze_graph' imported but unused](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/train.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  training_model = model.Model()
+  def model_fn(features, labels, mode):
+    training_model.set_mode(mode)
+    return training_model.get_model(features, labels)
+
+  return tf.estimator.Estimator(
+      model_fn=model_fn,
+      model_dir="model_output",
+      config=estimator_config)
+
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/train.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        for block_data in json_data:
+            action_type = block_data['type']
+            if action_type == 'MAIN_ACTION':
+                field = block_data['field']
+                data = self.read_field(field)
+                label = self.get_label(field['current_player_id'], player1_win)
+
+                all_data.append(data)
+                all_label.append(label)
+
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
+#!/usr/bin/python3
+
+import os
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: Codacy Issue found: [Unused import os](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/train.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/python3
+
+import argparse
+import numpy
+import random
+import os
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/train.py]: Codacy Issue found: [Unused import os](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/train.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+def main():
+  parser = argparse.ArgumentParser(description="Train Model")
+  parser.add_argument('-d', '--data', nargs='?', required=True, help="Data folder name")
+  parser.add_argument('--export', action='store_true', help="Export and freeze model")
+  try:
+    args = parser.parse_args()
+  except Exception:
+    parser.print_help()
+    sys.exit(0)
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/train.py]: Codacy Issue found: [undefined name 'sys'](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+    predictions = {
+        "classes": final_argmax,
+        "probabilities": tf.nn.softmax(final, name="softmax_tensor")}
+
+    if self._mode == tf.estimator.ModeKeys.PREDICT:
+      return tf.estimator.EstimatorSpec(
+          mode=self._mode, predictions=final,
+          export_outputs={"y": tf.estimator.export.PredictOutput({"y": final})})
+
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py (hunk 5 above)
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: Codacy Issue found: ['os' imported but unused](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/train.py (hunk 6 above)
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/train.py]: Codacy Issue found: ['os' imported but unused](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
+#!/usr/bin/python3
+
+import os
+import json
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: Codacy Issue found: ['json' imported but unused](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        self.data.extend([
+            0, 0, 0, 0,
+            -1.0, -1.0, -1.0, -1.0
+        ])
+
+    def add_minion_data(self, minion):
+        attackable = False
+        try:
+            attackable = minion['attackable']
+        except Exception as _:
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: Codacy Issue found: [local variable '_' is assigned to but never used](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        self.read_minions_data(board['opponent_player']['minions'])
+        self.add_current_hand(board['current_player']['hand'])
+        self.add_opponent_hand(board['opponent_player']['hand'])
+
+        self.add_mana_crystal(board['current_player']['mana_crystal'])
+        self.add_hero_power(board['current_player']['hero_power'])
+
+        return list(self.data)
+
+    def get_label(self, current_player, player1_win):
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+
+    for _ in range(data_reader.kCurrentHandCards):
+      outputs.extend(self._model_current_hand_card(input_getter))
+
+    return outputs
+
+  def _model_board_features(self, input_getter):
+    return [input_getter.get_rest()]
+
+  def _residual_block_unit(self, scope, inputs, hidden, add_input=None):
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+    outputs = []
+    outputs.append(
+        input_getter.get_next_slice(data_reader.kCurrentHandFeatures))
+
+    for _ in range(data_reader.kCurrentHandCards):
+      outputs.extend(self._model_current_hand_card(input_getter))
+
+    return outputs
+
+  def _model_board_features(self, input_getter):
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+        else:
+            assert False
+
+        assert player1_win is not None
+        if current_player_is_player1 is player1_win:
+            return LABEL_IF_PLAYER1_WIN
+        else:
+            return LABEL_IF_PLAYER1_LOSS
+
+    def win_or_loss(self, result):
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py (hunk 9 above)
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: Codacy Issue found: [Unused import json](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+    def read_hero_data(self, data):
+        health = data['hero']['health']
+        armor = data['hero']['armor']
+
+        self.data.extend([health + armor])
+
+    def add_minion_data_placeholder(self):
+        self.data.extend([
+            0, 0, 0, 0,
+            -1.0, -1.0, -1.0, -1.0
```
Comment (kozistr) [inline Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: 이 6차원 vector는 어떤 것을 의미하나요?

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+        ])
+
+    def add_minion_data(self, minion):
+        attackable = False
+        try:
+            attackable = minion['attackable']
+        except Exception as _:
+            attackable = False
```
Comment (kozistr) [inline Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: 위에 `attackable = False` 구문이 있으니 `pass` 사용해도 좋을 것 같습니다. (구문적인 것이니 다음 PR에 수정하도록 하겠습니다)

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+        def add_data(card_id, cost):
+            mapped_card_id = self.hand_card_mapper.get_index(card_id)
+            self.data.append(mapped_card_id)
+            self.data.append(cost)
+
+        for card in hand:
+            card_id = card['card_id']
+            cost = card['cost']
+            add_data(card_id, cost)
+        for _ in range(10 - len(hand)):
```
Comment (kozistr) [inline Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: `10` 같은 경우엔 scalability 를 고려해서 variable로 빼 두는게 좋을 것 같습니다.

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+    hero1 = tf.layers.dense(
+        name='hero1',
+        inputs=inputs,
+        units=self.kHeroConvolutionHidden1,
+        reuse=tf.AUTO_REUSE,
+        activation=tf.nn.relu)
+
+    return [hero1]
+
+  def _get_embedded_onboard_card_id(self, card_id):
```
Comment (kozistr) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: card_id 를 get하는 함수니 return type 이 str 이나 int로 통일이 되면 좋을것 같습니다.

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+import os
+import json
+import tensorflow as tf
+
+import data_reader
+
+
+kOutputNodeName = 'final_argmax'
+
+class NextInputGetter:
```
Comment (kozistr) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: indent 를 4 로 통일 해 주세요!

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+    final_argmax = tf.argmax(
+        name=kOutputNodeName,
+        input=final,
+        axis=1)
+
+    predictions = {
+        "classes": final_argmax,
+        "probabilities": tf.nn.softmax(final, name="softmax_tensor")}
+
+    if self._mode == tf.estimator.ModeKeys.PREDICT:
```
Comment (kozistr) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: 혹시 trainer baseline을 `tf.estimator` 를 사용한 이유가 있으신가용?!

File: Sources/Rosetta/Commons/JSONSerializer.cpp [hunk 21]
```diff
@@ -215,7 +215,7 @@ nlohmann::json JSONSerializer::SerializeSecret(const Spell& secret)
 {
     nlohmann::json obj;
 
-    obj["card_id"] = secret.card->id;
+    obj["card_id"] = secret.card->dbfID;
```
Comment (kozistr) [inline Sources/Rosetta/Commons/JSONSerializer.cpp]: `dbfID`는 어떤 형식의 data type, value를 가지나요?

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py  6
- Extensions/RosettaTraining/Scripts/TensorFlow/model.py  4
- Extensions/RosettaTraining/Scripts/TensorFlow/train.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929&bid=15844195)

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+    prev = dense1
+    for i in range(0, self.kResidualBlocks):
+      prev = self._residual_block(i+1, prev, self.kResidualBlockFeatures)
+
+    final = tf.layers.dense(
+        name='final',
+        inputs=prev,
+        units=1,
+        activation=None)
+
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+    for i in range(0, self.kResidualBlocks):
+      prev = self._residual_block(i+1, prev, self.kResidualBlockFeatures)
+
+    final = tf.layers.dense(
+        name='final',
+        inputs=prev,
+        units=1,
+        activation=None)
+		
+    final_argmax = tf.argmax(
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: Codacy Issue found: [local variable 'final_argmax' is assigned to but never used](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: Codacy Issue found: [Unused variable 'final_argmax'](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+            card_id = 0
+            cost = -1
+            add_data(card_id, cost)
+
+    def add_opponent_hand(self, hand):
+        if hand is None:
+            hand = {}
+
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/train.py [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+kTrainingPercent = 0.5
+kThreads = 20
+
+
+def _get_estimator():
+  sess_config = tf.ConfigProto(
+      intra_op_parallelism_threads=kThreads,
+      inter_op_parallelism_threads=kThreads)
+  estimator_config = tf.estimator.RunConfig(session_config=sess_config)
+
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/train.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+        self.data.extend([health + armor])
+
+    def add_minion_data_placeholder(self):
+        self.data.extend([
+            0, 0, 0, 0,
+            -1.0, -1.0, -1.0, -1.0
+        ])
+
+    def add_minion_data(self, minion):
+        attackable = False
```
Comment (JYPark09) [inline Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: 요 구문을 지우는 게 좋을 것 같아요.

File: Extensions/RosettaTraining/Scripts/TensorFlow/train.py [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+import tensorflow as tf
+from tensorflow.python.framework import graph_util, importer
+from tensorflow.python.platform import gfile
+from tensorflow.python.saved_model import loader, tag_constants
+from tensorflow.python.tools import saved_model_utils
+
+import data_reader
+import model
+
+kTrainingPercent = 0.5
```
Comment (JYPark09) [inline Extensions/RosettaTraining/Scripts/TensorFlow/train.py]: 학습 데이터셋의 비율을 더 높히는 게 좋을 것 같아요.

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py (hunk 15 above)
Comment (utilForever) [reply Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: 6차원이 아니라 8차원입니다. 이 벡터 데이터는 카드 ID, 체력, 최대 체력, 공격력, 공격 가능 여부, 도발 여부, 천상의 보호막 여부, 은신 여부를 저장합니다.

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py (hunk 16 above)
Comment (utilForever) [reply Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: `pass`로 수정합니다.

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py (hunk 17 above)
Comment (utilForever) [reply Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: 수정하겠습니다.

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py (hunk 18 above)
Comment (utilForever) [reply Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: card_id는 이제 string이 아닌 항상 int입니다.

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py (hunk 20 above)
Comment (utilForever) [reply Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: hearthstone-ai 코드를 참고했습니다.

File: Extensions/RosettaTraining/Scripts/TensorFlow/model.py (hunk 19 above)
Comment (utilForever) [reply Extensions/RosettaTraining/Scripts/TensorFlow/model.py]: 수정하겠습니다.

File: Sources/Rosetta/Commons/JSONSerializer.cpp (hunk 21 above)
Comment (utilForever) [reply Sources/Rosetta/Commons/JSONSerializer.cpp]: 하스스톤 내부 DB에서 사용하는 값으로 고유한 int 타입입니다.

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py (hunk 26 above)
Comment (utilForever) [reply Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: Exception 부분에 pass를 넣는 방향으로 수정했습니다.

File: Extensions/RosettaTraining/Scripts/TensorFlow/train.py (hunk 27 above)
Comment (utilForever) [reply Extensions/RosettaTraining/Scripts/TensorFlow/train.py]: 0.7로 올려두겠습니다.

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Extensions/RosettaTraining/Scripts/TensorFlow/model.py  4
- Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py  6
- Extensions/RosettaTraining/Scripts/TensorFlow/train.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929&bid=15844195)

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py [hunk 28]
```diff
@@ -0,0 +1,215 @@
...
+        try:
+            attackable = minion['attackable']
+        except Exception:
```
Comment (utilForever) [inline Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: Codacy Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929)

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py (hunk 26 above)
Comment (utilForever) [reply Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: 이렇게 하니 이슈가 발생하네요. 이 방향으로 수정하겠습니다.

File: Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py (hunk 16 above)
Comment (utilForever) [reply Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py]: 이렇게 하니 이슈가 발생하네요. `attackable = False` 문장을 고치는 방향으로 수정하겠습니다.

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Extensions/RosettaTraining/Scripts/TensorFlow/model.py  4
- Extensions/RosettaTraining/Scripts/TensorFlow/data_reader.py  6
- Extensions/RosettaTraining/Scripts/TensorFlow/train.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4748929&bid=15844195)

**Pull request merged**